packages/core/src/utils/debounce.ts

```diff
@@ -1,18 +1,37 @@
-type Fn<TA extends any[] = any[], TR = any> = (...args: TA) => TR;
-
 /**
- * Creates a debounced function that delays invoking `fun` until `ms` milliseconds
- * have passed since the last invocation of the debounced function.
- *
- * `fun` is invoked with the last arguments passed to the debounced function.
+ * Creates a debounced function that gets called at most once every "ms" milliseconds.
  *
- * Derived from `froebel/debounce`.
- * https://github.com/MathisBullinger/froebel/blob/main/debounce.ts
+ * 1) Function is being called for the first time, invoke immediately
+ * 2) Function has been called within the interval, but no timeout is set
+ * 3) Function has been called within the interval, timeout is already set
  */
-export function debounce<T extends Fn>(ms: number, fun: T) {
-  let toId: any;
-  return (...args: Parameters<T>) => {
-    clearTimeout(toId);
-    toId = setTimeout(() => fun(...args), ms);
+export function debounce<param extends unknown[], returnType>(
+  ms: number,
+  fun: (...x: param) => returnType,
+) {
+  let lastTimestamp = 0;
```

Comment:
Should this be initialized to `0`?

Reply:
Depends on whether or not the first call to debounce should have lag or not